### PR TITLE
refactor: finish canonical provider identifier audit

### LINKED ISSUE #960: [ENHANCEMENT] Replace remaining production provider literals and verify canonical identifier coverage
### Problem (one or two sentences)

After migrating the main provider consumers, isolated raw provider-name literals may remain in production code. These remaining comparisons could continue to bypass the canonical registry and leave the provider identifier centralization incomplete.

### Context (who is affected and when)

This affects maintainers and contributors during future provider additions, renames, and retirements. It is especially relevant after the targeted migrations, when remaining literals need to be classified as internal comparisons or intentional serialized, external, test, translation, documentation, model-name, or separate-domain values.

This is the final cleanup stage of the provider identifier centralization tracked by #944 and should follow the targeted consumer migrations.

### Desired behavior (conceptual, not technical)

All appropriate production comparisons, switch cases, and internal collections should use canonical provider identifiers. Intentional literals in persisted data, external formats, fixtures, translations, documentation, model IDs, and separate provider domains should remain readable and stable, with focused regression coverage confirming the completed migration.

### Constraints / preferences (optional)

Do not perform blind global replacement. Serialized values, external formats, fixtures, translations, documentation, model IDs, code-index provider identifiers, terminal provider identifiers, and other intentional literals should remain unchanged unless replacement provides a clear correctness or type-safety benefit.

### Request checklist

- [x] I've searched existing Issues and Discussions for duplicates
- [x] This describes a specific problem with clear context and impact

### Zoo Code Task Links (optional)

Parent tracking issue: #944

Foundational identifier registry: #951

Targeted migrations: #953, #954, #955, #956, #957, #958, and #959

### Acceptance criteria (optional)

- Remaining raw provider literals in production code are reviewed and classified.
- Appropriate internal API-provider comparisons, switch cases, and collections reference canonical identifiers.
- Intentional serialized, external, test, translation, documentation, model-name, and separate-domain literals remain unchanged.
- No API provider identifiers are confused with code-index, embedding, terminal, or other domain-specific provider names.
- Focused regression tests verify representative runtime usage.
- The acceptance criteria of #944 are reviewed and satisfied.
- Existing provider-related tests continue to pass with no intended user-facing behavior changes.

### Relevant code locations

This is a final audit rather than a predefined mechanical replacement. Re-scan production code after #953–#959, with particular attention to:

- [`packages/types/src`](https://github.com/Zoo-Code-Org/Zoo-Code/tree/main/packages/types/src) — remaining provider switches, comparisons, maps, and internal collections outside the completed foundational and default-model work.
- [`src/api`](https://github.com/Zoo-Code-Org/Zoo-Code/tree/main/src/api) — provider dispatch, capability checks, fetchers, and handler-specific comparisons not covered by #955 or #957.
- [`src/core`](https://github.com/Zoo-Code-Org/Zoo-Code/tree/main/src/core) — extension workflows not covered by #958.
- [`src/shared`](https://github.com/Zoo-Code-Org/Zoo-Code/tree/main/src/shared) — shared provider-dependent helpers not covered by #956.
- [`webview-ui/src`](https://github.com/Zoo-Code-Org/Zoo-Code/tree/main/webview-ui/src) — remaining UI provider comparisons not covered by #959.
- [`src/services/code-index/service-factory.ts`](https://github.com/Zoo-Code-Org/Zoo-Code/blob/main/src/services/code-index/service-factory.ts), [`src/shared/embeddingModels.ts`](https://github.com/Zoo-Code-Org/Zoo-Code/blob/main/src/shared/embeddingModels.ts), and [`webview-ui/src/components/chat/CodeIndexPopover.tsx`](https://github.com/Zoo-Code-Org/Zoo-Code/blob/main/webview-ui/src/components/chat/CodeIndexPopover.tsx) — classify carefully because these mix names shared with API providers and separate code-index identifiers such as `semble` and `openai-compatible`.
- [`src/integrations/terminal/TerminalRegistry.ts`](https://github.com/Zoo-Code-Org/Zoo-Code/blob/main/src/integrations/terminal/TerminalRegistry.ts) — exclude terminal provider names such as `vscode`; they are not API provider identifiers.
- Tests and fixtures across the repository — generally retain literal serialized values so compatibility tests continue to detect accidental string changes; replace only when the test is explicitly verifying canonical runtime usage.

### Proposed approach (optional)

Search production code for remaining active and retired provider strings after the targeted migrations. Review each occurrence individually, migrate internal API-provider identity comparisons, document intentional exceptions through surrounding context where necessary, and add focused regression coverage.

### Trade-offs / risks (optional)

Automated replacement could change external contracts, alter model IDs, reduce fixture readability, or conflate independent provider domains. This cleanup requires individual review of each occurrence and may uncover unrelated inconsistencies that should be reported separately rather than folded into this behavior-preserving refactor.

### PR DESCRIPTION
**Pull request by WebMad**
## Summary

- replace remaining raw API-provider comparisons in protocol selection with the canonical `providerIdentifiers` registry
- use canonical identifiers for configuration-free, Kimi Code OAuth, and Zoo Gateway session-auth readiness checks
- preserve serialized values and separate code-index, embedding, terminal, protocol, and model identifier domains
- add focused mutation-based regression coverage proving runtime consumers read the canonical registry

Closes #960

## TDD

Each migrated branch was driven by an observed red-green cycle:

- Vertex Claude protocol routing
- Vercel AI Gateway and Zoo Gateway Anthropic protocol routing
- OpenCode Go model-specific protocol routing
- configuration-free provider readiness
- Kimi Code OAuth readiness
- Zoo Gateway session-auth readiness

## Verification

- `cd packages/types && npx vitest run src/__tests__/provider-settings.test.ts src/__tests__/provider-identifiers.test.ts src/__tests__/provider-default-model.test.ts` — 38 passed
- `cd src && npx vitest run shared/__tests__/checkExistApiConfig.spec.ts shared/__tests__/ProfileValidator.spec.ts` — 50 passed
- types package type-check and changed-file ESLint — passed
- extension package type-check and changed-file ESLint — passed
- repository commit-time lint — passed

## Known unrelated environment issue

The repository pre-push type-check reaches the new Playwright visual-test files and fails because local dependencies `@playwright/experimental-ct-react`, `@playwright/test`, and `monocart-reporter` are unavailable. The types and extension packages affected by this PR type-check successfully. The local runtime is Node 24.7.0 while the repository requests Node 22.23.1.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved API protocol routing for Anthropic-style vs OpenAI-style models, including Vertex, Zoo Gateway/Vercel AI Gateway canonical identifiers, and Opencode Go formatting.
  * Improved reliability of provider detection for authentication/config checks by using canonical provider identifiers, covering keyless, OAuth (including Kimi Code), and session-based scenarios.
  * Strengthened edge-case behavior for undefined/empty providers and case-insensitive Claude detection.
* **Tests**
  * Updated and expanded test coverage to use shared protocol constants and canonical provider identifiers, adding explicit wire-string assertions and Gateway provider routing cases.

### CODE REVIEW

File: packages/types/src/__tests__/provider-settings.test.ts [hunk 1]
```diff
@@ -15,6 +15,19 @@ describe("getApiProtocol", () => {
 	})
 
 	describe("Vertex provider with Claude models", () => {
+		it("uses the canonical Vertex identifier for Claude protocol selection", () => {
+			const identifiers = providerIdentifiers as Record<string, string>
+			const originalIdentifier = identifiers.vertex!
```
Comment (WebMad) [inline packages/types/src/__tests__/provider-settings.test.ts]: I don’t think it’s appropriate to change values in identifiers. I also don’t think we should use the bang operator (`!`).

Comment (WebMad) [reply packages/types/src/__tests__/provider-settings.test.ts]: The same implementation appears in this file as well. Please pay attention to these cases too

File: packages/types/src/provider-settings.ts [hunk 2]
```diff
@@ -581,21 +581,25 @@ export const modelIdKeysByProvider: Record<TypicalProvider, ModelIdKey> = {
...
+	providerIdentifiers.minimax,
+]
 
 export const getApiProtocol = (provider: ProviderName | undefined, modelId?: string): "anthropic" | "openai" => {
 	if (provider && ANTHROPIC_STYLE_PROVIDERS.includes(provider)) {
 		return "anthropic"
 	}
 
-	if (provider && provider === "vertex" && modelId && modelId.toLowerCase().includes("claude")) {
+	if (provider && provider === providerIdentifiers.vertex && modelId && modelId.toLowerCase().includes("claude")) {
```
Comment (WebMad) [inline packages/types/src/provider-settings.ts]: I don’t like that the "claude" string is still hardcoded here. Could we extract it into a constant?

File: packages/types/src/provider-settings.ts [hunk 3]
```diff
@@ -581,21 +581,25 @@ export const modelIdKeysByProvider: Record<TypicalProvider, ModelIdKey> = {
...
 		return "anthropic"
```
Comment (WebMad) [inline packages/types/src/provider-settings.ts]: The same issue applies here. I think this could be extracted into a constant as well

File: packages/types/src/provider-settings.ts [hunk 4]
```diff
@@ -581,21 +581,25 @@ export const modelIdKeysByProvider: Record<TypicalProvider, ModelIdKey> = {
...
 	}
 
 	// Vercel AI Gateway and Zoo Gateway use the anthropic protocol for anthropic models.
 	if (
 		provider &&
-		["vercel-ai-gateway", "zoo-gateway"].includes(provider) &&
+		(provider === providerIdentifiers.vercelAiGateway || provider === providerIdentifiers.zooGateway) &&
```
Comment (WebMad) [inline packages/types/src/provider-settings.ts]: I think the previous implementation was clearer. Could we preserve the original condition structure in the `if` statement?

File: packages/types/src/__tests__/provider-settings.test.ts (hunk 1 above)
Comment (WebMad) [reply packages/types/src/__tests__/provider-settings.test.ts]: Addressed in 131c1e103. I removed the mutation-based providerIdentifiers tests, including the Record casts and non-null assertions. The canonical registry values are now passed directly in the existing behavioral tests. I applied the same cleanup to the Vertex, gateway, OpenCode Go, configuration-free, Kimi Code, and Zoo Gateway cases.

File: packages/types/src/provider-settings.ts (hunk 2 above)
Comment (WebMad) [reply packages/types/src/provider-settings.ts]: Addressed in 131c1e103. The hardcoded Claude fragment is now the CLAUDE_MODEL_ID_FRAGMENT constant and the protocol condition references that constant.

File: packages/types/src/provider-settings.ts (hunk 3 above)
Comment (WebMad) [reply packages/types/src/provider-settings.ts]: Addressed in 131c1e103. The hardcoded Anthropic model prefix is now the ANTHROPIC_MODEL_ID_PREFIX constant.

File: packages/types/src/provider-settings.ts (hunk 4 above)
Comment (WebMad) [reply packages/types/src/provider-settings.ts]: Addressed in 131c1e103. I restored the original array-includes condition structure by introducing ANTHROPIC_MODEL_GATEWAY_PROVIDERS and checking ANTHROPIC_MODEL_GATEWAY_PROVIDERS.includes(provider).

File: packages/types/src/provider-settings.ts [hunk 5]
```diff
@@ -581,23 +581,39 @@ export const modelIdKeysByProvider: Record<TypicalProvider, ModelIdKey> = {
...
 	// Vercel AI Gateway and Zoo Gateway use the anthropic protocol for anthropic models.
 	if (
 		provider &&
-		["vercel-ai-gateway", "zoo-gateway"].includes(provider) &&
+		ANTHROPIC_MODEL_GATEWAY_PROVIDERS.includes(provider) &&
 		modelId &&
-		modelId.toLowerCase().startsWith("anthropic/")
+		modelId.toLowerCase().startsWith(ANTHROPIC_MODEL_ID_PREFIX)
 	) {
 		return "anthropic"
```
Comment (WebMad) [inline packages/types/src/provider-settings.ts]: The `"antropic"` string should also be extracted into a constant.

Comment (WebMad) [reply packages/types/src/provider-settings.ts]: Addressed in 2ff97ad29. I extracted the repeated Anthropic protocol return value into ANTHROPIC_API_PROTOCOL and now use it for every Anthropic branch in getApiProtocol.

File: packages/types/src/__tests__/provider-settings.test.ts [hunk 6]
```diff
@@ -16,7 +16,7 @@ describe("getApiProtocol", () => {
 
 	describe("Vertex provider with Claude models", () => {
 		it("should return 'anthropic' for vertex provider with claude models", () => {
-			expect(getApiProtocol("vertex", "claude-3-opus")).toBe("anthropic")
+			expect(getApiProtocol(providerIdentifiers.vertex, "claude-3-opus")).toBe("anthropic")
```
Comment (WebMad) [inline packages/types/src/__tests__/provider-settings.test.ts]: Why is `providerIdentifiers.vertex` used only here? I can see the same hardcoded strings on the following lines, but they haven’t been replaced.

File: packages/types/src/__tests__/provider-settings.test.ts [hunk 7]
```diff
@@ -34,6 +34,11 @@ describe("getApiProtocol", () => {
 	})
 
 	describe("Vercel AI Gateway provider", () => {
+		it("uses canonical gateway identifiers for Anthropic model protocol selection", () => {
+			expect(getApiProtocol(providerIdentifiers.vercelAiGateway, "anthropic/claude-3-opus")).toBe("anthropic")
```
Comment (WebMad) [inline packages/types/src/__tests__/provider-settings.test.ts]: Could we use the `ANTHROPIC_API_PROTOCOL` constant here and in similar cases?

File: packages/types/src/__tests__/provider-settings.test.ts [hunk 8]
```diff
@@ -55,7 +60,7 @@ describe("getApiProtocol", () => {
 
 	describe("Opencode Go provider", () => {
 		it("should return 'anthropic' for opencode-go Anthropic-format models (Qwen/MiniMax)", () => {
-			expect(getApiProtocol("opencode-go", "qwen3.7-max")).toBe("anthropic")
+			expect(getApiProtocol(providerIdentifiers.opencodeGo, "qwen3.7-max")).toBe("anthropic")
 			expect(getApiProtocol("opencode-go", "qwen3.7-plus")).toBe("anthropic")
```
Comment (WebMad) [inline packages/types/src/__tests__/provider-settings.test.ts]: I think we could use `providerIdentifiers.opencodeGo` on the following lines as well

File: packages/types/src/provider-settings.ts [hunk 9]
```diff
@@ -581,25 +581,42 @@ export const modelIdKeysByProvider: Record<TypicalProvider, ModelIdKey> = {
...
-export const ANTHROPIC_STYLE_PROVIDERS: ProviderName[] = ["anthropic", "bedrock", "minimax"]
+export const ANTHROPIC_STYLE_PROVIDERS: ProviderName[] = [
+	providerIdentifiers.anthropic,
+	providerIdentifiers.bedrock,
+	providerIdentifiers.minimax,
+]
+
+const CLAUDE_MODEL_ID_FRAGMENT = "claude"
+const ANTHROPIC_MODEL_ID_PREFIX = "anthropic/"
+const ANTHROPIC_API_PROTOCOL = "anthropic"
```
Comment (WebMad) [inline packages/types/src/provider-settings.ts]: Is this the right place to define constants related to Anthropic?

File: packages/types/src/__tests__/provider-settings.test.ts (hunk 6 above)
Comment (WebMad) [reply packages/types/src/__tests__/provider-settings.test.ts]: Addressed in abb5388bb. Every Vertex case in this test suite now uses providerIdentifiers.vertex, including Claude, non-Claude, empty-model, no-model, and case-insensitivity scenarios.

File: packages/types/src/__tests__/provider-settings.test.ts (hunk 7 above)
Comment (WebMad) [reply packages/types/src/__tests__/provider-settings.test.ts]: Addressed in abb5388bb. Anthropic protocol assertions now use ANTHROPIC_API_PROTOCOL throughout the provider settings test suite instead of repeating the literal.

File: packages/types/src/__tests__/provider-settings.test.ts (hunk 8 above)
Comment (WebMad) [reply packages/types/src/__tests__/provider-settings.test.ts]: Addressed in abb5388bb. All OpenCode Go scenarios now use providerIdentifiers.opencodeGo, including Anthropic-format, OpenAI-format, missing-model, and unknown-model cases.

File: packages/types/src/provider-settings.ts (hunk 9 above)
Comment (WebMad) [reply packages/types/src/provider-settings.ts]: Addressed in abb5388bb. I moved ANTHROPIC_API_PROTOCOL, ANTHROPIC_MODEL_ID_PREFIX, and CLAUDE_MODEL_ID_FRAGMENT to providers/anthropic.ts, then import them into provider-settings.ts through the providers barrel.

File: packages/types/src/__tests__/provider-settings.test.ts [hunk 10]
```diff
@@ -1,90 +1,107 @@
...
+			expect(getApiProtocol(providerIdentifiers.vertex, "Claude-3-Sonnet")).toBe(ANTHROPIC_API_PROTOCOL)
+			expect(getApiProtocol(providerIdentifiers.vertex, "CLAUDE-instant")).toBe(ANTHROPIC_API_PROTOCOL)
+			expect(getApiProtocol(providerIdentifiers.vertex, "anthropic/claude-3-haiku")).toBe(ANTHROPIC_API_PROTOCOL)
 		})
 
 		it("should return 'openai' for vertex provider with non-claude models", () => {
-			expect(getApiProtocol("vertex", "gpt-4")).toBe("openai")
-			expect(getApiProtocol("vertex", "gemini-pro")).toBe("openai")
-			expect(getApiProtocol("vertex", "llama-2")).toBe("openai")
+			expect(getApiProtocol(providerIdentifiers.vertex, "gpt-4")).toBe("openai")
```
Comment (WebMad) [inline packages/types/src/__tests__/provider-settings.test.ts]: Could we replace the `"openai"` string with a constant?

Comment (WebMad) [reply packages/types/src/__tests__/provider-settings.test.ts]: Addressed in 14f7277b3. I added OPENAI_API_PROTOCOL to providers/openai.ts, use it as the getApiProtocol fallback, and replaced all repeated OpenAI protocol literals in the provider settings tests.

File: packages/types/src/__tests__/provider-settings.test.ts [hunk 11]
```diff
@@ -1,107 +1,124 @@
...
 		})
 
 		it("should return 'anthropic' for bedrock provider", () => {
-			expect(getApiProtocol("bedrock")).toBe("anthropic")
-			expect(getApiProtocol("bedrock", "gpt-4")).toBe("anthropic")
-			expect(getApiProtocol("bedrock", "claude-3-opus")).toBe("anthropic")
+			expect(getApiProtocol(providerIdentifiers.bedrock)).toBe(ANTHROPIC_API_PROTOCOL)
+			expect(getApiProtocol(providerIdentifiers.bedrock, "gpt-4")).toBe(ANTHROPIC_API_PROTOCOL)
+			expect(getApiProtocol(providerIdentifiers.bedrock, "claude-3-opus")).toBe(ANTHROPIC_API_PROTOCOL)
 		})
```
Comment (edelauna) [inline packages/types/src/__tests__/provider-settings.test.ts]: `minimax` is in `ANTHROPIC_STYLE_PROVIDERS` alongside `anthropic` and `bedrock`, but there's no test calling `getApiProtocol(providerIdentifiers.minimax, ...)`. Could we add a case to this block so a mutation removing `minimax` from the list would be caught?

File: packages/types/src/__tests__/provider-settings.test.ts [hunk 12]
```diff
@@ -1,107 +1,124 @@
-import { getApiProtocol } from "../provider-settings.js"
+import { ANTHROPIC_API_PROTOCOL, getApiProtocol, OPENAI_API_PROTOCOL, providerIdentifiers } from "../index.js"
```
Comment (edelauna) [inline packages/types/src/__tests__/provider-settings.test.ts:1]: Every protocol assertion here uses the imported constant as the expected value (e.g. `.toBe(ANTHROPIC_API_PROTOCOL)`), so the tests verify internal consistency but don't pin the wire value `"anthropic"`. If both the constant and its use site agreed on a wrong value, all tests would still pass. Worth adding a small companion test that asserts `ANTHROPIC_API_PROTOCOL === "anthropic"` and `OPENAI_API_PROTOCOL === "openai"` to anchor the wire values?

File: packages/types/src/__tests__/provider-settings.test.ts [hunk 13]
```diff
@@ -1,107 +1,124 @@
...
 
 		it("should return 'openai' for vertex provider without model", () => {
-			expect(getApiProtocol("vertex")).toBe("openai")
+			expect(getApiProtocol(providerIdentifiers.vertex)).toBe(OPENAI_API_PROTOCOL)
 		})
 	})
 
 	describe("Vercel AI Gateway provider", () => {
+		it("uses canonical gateway identifiers for Anthropic model protocol selection", () => {
+			expect(getApiProtocol(providerIdentifiers.vercelAiGateway, "anthropic/claude-3-opus")).toBe(
```
Comment (edelauna) [inline packages/types/src/__tests__/provider-settings.test.ts]: The `vercelAiGateway` call here is identical to the first assertion in the `"should return 'anthropic' for vercel-ai-gateway..."` test at line 47 — is `zooGateway` (line 41) the only new coverage this test adds? If so, trimming the duplicate would make the intent clearer.

File: packages/types/src/providers/anthropic.ts [hunk 14]
```diff
@@ -5,6 +5,9 @@ import type { ModelInfo } from "../model.js"
 
 export type AnthropicModelId = keyof typeof anthropicModels
 export const anthropicDefaultModelId: AnthropicModelId = "claude-sonnet-4-5"
+export const ANTHROPIC_API_PROTOCOL = "anthropic"
+export const ANTHROPIC_MODEL_ID_PREFIX = "anthropic/"
+export const CLAUDE_MODEL_ID_FRAGMENT = "claude"
```
Comment (edelauna) [inline packages/types/src/providers/anthropic.ts]: `ANTHROPIC_MODEL_ID_PREFIX` and `CLAUDE_MODEL_ID_FRAGMENT` are each used exactly once in `provider-settings.ts` and have no callers elsewhere in `src/`. Is there a reason to export them as public `@roo-code/types` surface rather than keeping them unexported — like `ANTHROPIC_MODEL_GATEWAY_PROVIDERS` on line 594 of `provider-settings.ts`?

**Review by edelauna — CHANGES_REQUESTED** [review]

File: packages/types/src/__tests__/provider-settings.test.ts (hunk 11 above)
Comment (WebMad) [reply packages/types/src/__tests__/provider-settings.test.ts]: Addressed in fcd904d65. I replaced the separate Anthropic and Bedrock cases with an independent parameterized matrix covering anthropic, bedrock, and minimax. The test data intentionally does not derive from ANTHROPIC_STYLE_PROVIDERS, so removing minimax from the production list is caught.

File: packages/types/src/__tests__/provider-settings.test.ts (hunk 12 above)
Comment (WebMad) [reply packages/types/src/__tests__/provider-settings.test.ts:1]: Addressed in fcd904d65. Added a companion assertion that pins ANTHROPIC_API_PROTOCOL to "anthropic" and OPENAI_API_PROTOCOL to "openai", while the routing assertions continue to use the constants.

File: packages/types/src/__tests__/provider-settings.test.ts (hunk 13 above)
Comment (WebMad) [reply packages/types/src/__tests__/provider-settings.test.ts]: Addressed in fcd904d65. Removed the duplicate Vercel assertion and narrowed that test to the Zoo Gateway canonical identifier. I also renamed the enclosing describe block to "Gateway providers" so it accurately covers both gateway-specific test groups.

File: packages/types/src/providers/anthropic.ts (hunk 14 above)
Comment (WebMad) [reply packages/types/src/providers/anthropic.ts]: Addressed in fcd904d65. Removed both values from the exported provider module and kept them as unexported constants next to ANTHROPIC_MODEL_GATEWAY_PROVIDERS in provider-settings.ts.

**Review by edelauna — APPROVED** [review]
Thank you
